### PR TITLE
Abstract away jedis using RedisPublisher

### PR DESCRIPTION
**Pull request by drosowski**
Using Spring Boot 3.x, Jedis is no longer the default implementation. To support the users choice of redis client, this PR will remove the direct dependency to jedis. Instead we inverse the dependency by providing the `RedisPublisher` interface, that should be implemented by the user of the emitter.
**QUESTION:** Testing it I wrapped Springs RedisConnection in a class implementing RedisPublisher. But this way we only use one connection. What happens if the connection fails? It won't reconnect in this case I guess.

### CODE REVIEW

File: src/main/kotlin/de/smartsquare/socketio/emitter/RedisPublisher.kt [hunk 1]
```diff
@@ -0,0 +1,5 @@
+package de.smartsquare.socketio.emitter
+
+interface RedisPublisher {
+    fun publish(channel: ByteArray, message: ByteArray)
```
Comment (rubengees) [inline src/main/kotlin/de/smartsquare/socketio/emitter/RedisPublisher.kt]: I think `String` might be a better type for the channel.

File: build.gradle [hunk 2]
```diff
@@ -36,15 +35,14 @@ plugins {
 }
 
 group = "de.smartsquare"
-version = System.getenv("GITHUB_VERSION") ?: "1.0.0-SNAPSHOT"
+version = System.getenv("GITHUB_VERSION") ?: "1.0.1-SNAPSHOT"
```
Comment (rubengees) [inline build.gradle]: This version is only for local development and does not need to be incremented.

**Review by rubengees — CHANGES_REQUESTED** [review]
Cool!

To make this more plug and play, I would like to have a default implmentation for the two popular redis clients though. We have 2 options:

- Make two separate gradle modules `socket-io-redis-emitter-jedis` and `socket-io-redis-emitter-lettuce` with each framework as dependency and the default implementation for the `RedisPublisher`.
- Depend on both frameworks via `compileOnly` and rely on the user to depend on the framework which is used.

1 is more correct, 2 is (much) simpler.

**Comment by rubengees** [comment]
Not tested but the default implementation for both could look like this:

```kt
class JedisPublisher(private val jedis: JedisPool): RedisPublisher {
    override fun publish(channel: String, message: ByteArray) {
        jedis.resource.use { it.publish(channel.toByteArray(), message) }
    }
}

class LettucePublisher(private val connection: StatefulRedisConnection<Any, Any>): RedisPublisher {
    override fun publish(channel: String, message: ByteArray) {
        connection.use { it.sync().publish(channel, message.decodeToString()) }
    }
}
```

File: README.md [hunk 3]
```diff
@@ -38,14 +38,18 @@ repositories {
 }
 
 dependencies {
-    implementation "de.smartsquare:socket-io-redis-emitter:0.11.3"
+    implementation "de.smartsquare:socket-io-redis-emitter:0.11.3"
```
Comment (rubengees) [inline README.md]: GitHub shows some unrelated change here, please revert :slightly_smiling_face:

File: README.md [hunk 4]
```diff
@@ -38,14 +38,18 @@ repositories {
...
 
+This library comes with a default implementation for both jedis and lettuce.
+
 ### Emit Cheatsheet
 
+When using Jedis:
 ```kotlin
-val emitter = Emitter(JedisPool("localhost"), namespace = "/")
+val jedisConnection = jedisConnectionFactory.connection as JedisConnection
+val emitter = Emitter(JedisPublisher(jedisConnection.jedis))
```
Comment (rubengees) [inline README.md]: Is this Spring specific code? I don't think this belongs in the README (or would need a specific section) since this is a package without dependency on Spring.

File: build.gradle [hunk 5]
```diff
@@ -36,15 +37,14 @@ plugins {
...
```
Comment (rubengees) [inline build.gradle]: Revert please :slightly_smiling_face:

**Review by rubengees — CHANGES_REQUESTED** [review]
:+1:

File: README.md (hunk 4 above)
Comment (drosowski) [reply README.md]: I added a dedicated section for this.

**Comment by drosowski** [comment]
The `example` is not running anymore. When depending on the client to provide the redis client implementation, how can we build a runnable example?

**Review by rubengees — APPROVED** [review]
Looks good :ok_hand:
